specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+|    4     | subnet size (1 byte) | raw_addr size (1 byte) | subnetID (up to 74 bytes) | separator (1 byte) | raw address (up to 66 bytes) | 
+|----------|-------------------------------------------------------------------------------------------------------------------------------|
+```
+
+With hierarchical addresses three new functions are introduced to Filecoin addresses:
+- `Raw()`: Returns the raw address of the hierarchical address. If it is not an HA, it still returns the corresponding raw address.
+- `Subnet()`: It returns the subnet ID of the HA. It returns an error or an `Subnet.Undef` if the address is not an HA.
+- `Levels()`: Returns the number of levels in the `SubnetID` included in the HA address.
+- `PrettyString()`: Returns a beautified version of an HA address like `/root/f0100:<raw address string>`. This is method is only available for HA. For the rest of addresses it just returns `String()`. 
+
+Finally, a pair of keys from a user control the same address in every subnet in the system. Thus, the raw address determines the ownership of a specific HA. 
+
+## Checkpointing
+Checkpoints are used to anchor a subnet's security to that of its parent network, as well as to propagate information from a child chain to other subnets in the system. Checkpoints for a subnet can be verified at any point using the state of the subnet chain which can then be used to generate proofs of misbehaviors in the subnet (or so-called _fraud/fault proofs_) which, in turn, can be used for penalizing misbehaving entities (_"slashing"_). See [detectable misbehaviors](#Detectable-misbehaviors "Detectable misbehaviors") for further details.
+
+Checkpoints need to be signed by miners of a child chain and committed to the parent chain through their corresponding `SA`. The specific signature policy is defined in the `SA` and determines the type and minimum number of signatures required for a checkpoint to be accepted and validated by the `SA` for its propagation to the top chain. Different signature schemes may be used here, including multi-signatures or threshold signatures among subnet miners. For instance, in the reference implementation of `SA`, the actor waits for more than `2/3` of the validators in the subnet to send the valid checkpoint signed before propagating it to the `SCA` for commitment. Alternative verification policies for a subnet can be implemented in the `SubmitCheckpoint()` function of the subnet actor interface. 
+
+In order for a new checkpoint to be accepted for commitment in `SCA`, the source of the message `CommitChildCheckpoint()` needs to be the address of the subnet actor of the corresponding subnet; the subnet needs to be in an `Active` state (i.e. its collateral is over `CollateralThreshold`); the epoch of the checkpoint should be a multiple of `CheckPeriod` and larger that that of the previous checkpoint; and the checkpoint must point to the `CID` of the previous checkpoint committed by the subnet. 
```

Comment:
> In order for a new checkpoint to be accepted for commitment in `SCA` ... the epoch of the checkpoint should be a multiple of `CheckPeriod` and larger that that of the previous checkpoint;

I found this to be a bit confusing. It sounds SCA here is the one in the parent subnet, and it's just received a checkpoint created by the child SCA in the child subnet, submitted through the SA that lives in the parent subnet. 

Now, I understand that the SCA in the child subnet creates checkpoints at each `CheckPeriod`, but I don't see the check period of the child being tracked by the parent, e.g. it's not part of the `Subnet` struct. 

Surely we don't expect all child checkpoints to come at the period of the parent, as that would contradict the next sentence, which states that they can set it according to the needs of their consensus algorithms. 

So, it sounds to me that if the SA is happy with the checkpoint, the SCA should accept it (if the subnet is active and the checkpoint points at the previous one).

Reply:
> So, it sounds to me that if the SA is happy with the checkpoint, the SCA should accept it (if the subnet is active and the checkpoint points at the previous one).

That is correct. The SCA does some basic semantic validations at this point (with access to the subnet state there's not much we can do). Let's see if we can start exploring Lurk soon, as that would improve this flow a lot.